### PR TITLE
[WIP][Feature] Migrate to MR v2

### PR DESCRIPTION
**Pull request by tzhouam**
## Purpose
This PR aims to migrate to model runner v2
## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Review by lishunyang12 — COMMENTED** [review]
## Early Architectural Review (WIP)

Reviewed the full diff (~5700 additions). This is a substantial migration from MR V1 to MR V2 for Omni model stages. The design document is thorough and the layered architecture (OmniGPUModelRunner -> OmniARModelRunner, OmniModelState + plugins, OmniIntermediateBuffer) is well-motivated. Below are observations and concerns for the authors to address before moving out of WIP.

---

### Architecture & Design

1. **`_OMNI_ARCHITECTURES` mismatch between source and test.** The source set in `vllm_omni/worker_v2/model_states/__init__.py` includes `"Qwen2_5OmniForConditionalGeneration"` but the test `test_omni_architectures_set_contains_expected` does NOT include it — the test asserts a strict equality check and will fail. Either add the missing arch to the test or remove it from the source set. This is a real bug.

2. **Monkey-patching `get_rope_state` in `OmniModelState.__init__`** via a class-level lock + temporary override of `_default_mod.get_rope_state` is fragile. If upstream renames or moves `get_rope_state`, this breaks silently. Consider adding a comment pointing to the specific upstream location being patched, and ideally proposing an upstream hook (e.g., a `try_get_rope_state` that returns `None` instead of asserting) in a follow-up issue.

3. **Monkey-patching `init_model_state` in `OmniGPUModelRunner.load_model`** (lines patching `_mr_module.init_model_state = init_omni_model_state`) has the same fragility. This is a recurring pattern — three separate monkey-patches in OmniModelState.__init__ and OmniGPUModelRunner.load_model. Consider documenting these patches in a central location (e.g., a `PATCHES.md` or inline docstring listing all upstream functions being overridden).

4. **`OmniGPUModelRunner.execute_model` re-implements most of `GPUModelRunner.execute_model`** rather than calling `super().execute_model()` as the design doc proposed. The ~150-line override duplicates batch descriptor dispatch, DP sync, input preparation, attention metadata, MM embeddings, model forward, and KV connector calls. This means any upstream changes to `GPUModelRunner.execute_model` must be manually merged. This is the highest-risk part of the migration. Consider whether pre/post hooks or wrapping (as the design doc originally proposed) could reduce this duplication.

5. **`OmniGenerationModelRunner.execute_model` also duplicates significant logic** from the upstream `GPUModelRunner`. Between `omni_model_runner.py` and `omni_generation_model_runner.py`, there are two near-complete copies of the upstream execute_model flow. This is ~600 lines of duplicated infrastructure code that will drift from upstream.

6. **`compile_or_warm_up_model` in `GPUARWorker`** patches `warmup_kernels` by replacing the function in two separate modules (`vllm.v1.worker.gpu_worker` and `vllm.v1.worker.gpu.warmup`). The `_noop` function returns `None` explicitly — it should just use `pass` or `return`. Minor, but worth cleaning up.

### Correctness Concerns

7. **`prepare_attn` override uses actual `max_seq_len` instead of `max_model_len`.** The comment explains this is needed to match V1 MR behavior and avoid FlashAttention tiling differences. However, this means the attention kernel may encounter sequences longer than what it was configured for during CUDA graph capture (where `for_capture=True` uses `max_model_len`). Is there a risk of runtime mismatch between capture and replay?

8. **`_handle_async_chunk_updates` removes and re-adds requests to `req_states`.** The sequence is: `req_states.remove_request(req_id)` then `req_states.add_request(req_id, ...)`. If the slot freed by `remove_request` is immediately reused by another concurrent operation (unlikely in single-threaded scheduler, but worth noting), the re-add could get a different slot index. The comment says this is safe for Code2Wav, but the safety argument should be more explicit about single-threaded execution guarantees.

9. **`OmniAsyncOutput` reimplements async D2H copy logic** rather than extending the upstream `AsyncOutput`. If upstream changes the copy pattern (e.g., adds new fields, changes stream synchronization), this will silently diverge. Consider inheriting from or delegating to the upstream `AsyncOutput` where possible.

10. **`chunk_transfer_adapter.py` change: `request._all_token_ids = list(new_ids)`** — this directly modifies a private attribute on the `Request` object. If the upstream `Request` class changes how `_all_token_ids` is managed (e.g., becomes a property or uses a different internal representation), this will break. Add a comment noting this coupling.

### Test Coverage

11. **Good unit test structure.** The `tests/worker_v2/` suite covers the key components (intermediate buffer, model state, plugin, runners, factory). The use of `object.__new__()` to skip real constructors is a pragmatic pattern for unit testing.

12. **Missing integration tests.** The design doc's Phase 2/3 checklist mentions integration tests (M-RoPE consistency, pooler_output consistency, KV transfer), but none are included in this PR. These are critical for correctness validation.

13. **`test_generation_scheduler_finish_condition.py` tests a re-implemented finish condition** rather than the actual scheduler code. The `_evaluate_finish_condition` function in the test file is a manual extraction of the logic — if the real scheduler logic changes, this test won't catch the regression. Consider testing the actual `update_from_output` method instead.

### Code Quality

14. **Design document is in Chinese.** The `docs/design/feature/qwen3_omni_thinker_mr_v2.md` is 1153 lines entirely in Chinese. For an open-source project with international contributors, consider adding at least an English summary section, or translating the key sections.

15. **`end2end_benchmark_script.py` (640 lines)** monkey-patches `omni._log_summary_and_cleanup` via `_SummaryCollector`. This relies on a private method and the internal `request_states` dict. The code itself acknowledges this fragility. Consider whether the benchmark needs to ship in the same PR as the core migration.

16. **Shell test scripts (`run_tests.sh`, `run_tests_async_chunk.sh`)** hard-code model paths like `/models/Qwen3-Omni-30B-A3B-Instruct`. These should be parameterized or documented as requiring local setup.

17. **`torch.accelerator.*` → `torch.cuda.*` replacements** in `gpu_ar_worker.py` (e.g., `torch.accelerator.device_count()` → `torch.cuda.device_count()`, `torch.accelerator.set_device_index()` → `current_platform.set_device()`). These changes reduce platform generality. Was `torch.accelerator` intentionally used for multi-platform support? If so, this regression should be reverted; if the project is CUDA-only, this is fine but should be noted.

18. **Minor: `qwen3_omni_moe_thinker.py` introduces a typo** at line 1265: `"correspoending"` should be `"corresponding"`.

### Summary

The overall architecture is sound — the `ModelState` + plugin pattern is a good fit for Omni's multi-stage pipeline, and the `OmniIntermediateBuffer` with req_index-based O(1) access is a clear improvement over V1's dict-based approach. The main risks are:

- **Maintenance burden from duplicated `execute_model` logic** (~600 lines across two runners that mirror upstream)
- **Three monkey-patches** that couple tightly to upstream internals
- **Missing integration tests** for the correctness-critical paths (M-RoPE, pooler_output, KV transfer)

These should be addressed before this moves out of WIP.

File: vllm_omni/worker_v2/model_states/plugin.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+from abc import ABC, abstractmethod
+from typing import Any
+
+from vllm.v1.core.sched.output import NewRequestData
+from vllm.v1.worker.gpu.input_batch import InputBatch
+from vllm.v1.worker.gpu.states import RequestState
+
+
+class OmniModelStatePlugin(ABC):
+    """Extension point for model-specific behavior in OmniModelState."""
```
Comment (linyueqian) [inline vllm_omni/worker_v2/model_states/plugin.py:20]: This plugin surface covers request lifecycle, inputs, and postprocess, but the first real tenant from the TTS side will need two hooks it does not have yet: Fish Speech's kvcache attention backend. In the v1 runner it is wired by direct import at two generic call sites (`gpu_model_runner.py`: `prewarm_fish_kvcache_attn_capture_workspaces` before CUDA graph capture, and `maybe_attach_fish_kvcache_seq_lens_upper_bound` on every attention-metadata build), which is exactly the pattern #4855 asks to clean up. Suggest adding two optional no-op hooks here so Fish migrates through the plugin instead of re-importing itself into the v2 runner:

```python
def prewarm_capture(self, *, model_config, device, dtype, capture_sizes) -> None:
    pass

def attach_attn_metadata(self, *, attn_metadata, input_batch, seq_lens_cpu,
                         num_reqs, num_reqs_padded, max_query_len,
                         pad_attn, for_cudagraph_capture) -> None:
    pass
```

The kwarg surface is lifted from the two v1 call sites, and every argument is a generic scheduling fact rather than a fish concept, so nothing model-specific leaks into the interface. Cheap for everyone else: the runner iterates an (almost always empty) plugin list it resolved once at load.

File: vllm_omni/worker_v2/model_states/omni_model_state.py [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+        self._mtp_hidden: torch.Tensor | None = None
+        self._mtp_text_step: torch.Tensor | None = None
+        if self._embed_dim > 0 and hasattr(model, "talker_mtp"):
+            max_bs = max_num_reqs
+            self._mtp_input_ids = torch.zeros(max_bs, dtype=torch.long, device=device)
+            self._mtp_input_embeds = torch.zeros((max_bs, self._embed_dim), dtype=self.dtype, device=device)
+            self._mtp_hidden = torch.zeros((max_bs, self._embed_dim), dtype=self.dtype, device=device)
+            self._mtp_text_step = torch.zeros((max_bs, self._embed_dim), dtype=self.dtype, device=device)
+
+        if hasattr(model, "get_omni_plugins"):
```
Comment (linyueqian) [inline vllm_omni/worker_v2/model_states/omni_model_state.py:156]: `get_omni_plugins()` on the model instance is the right scoping for the multi-stage world (a global registry would make, say, a stage-1 Code2Wav runner iterate a stage-0 model's extensions). One convention worth documenting here: plugins should register only when their feature actually engaged, not merely when enabled. For the Fish case that means `install_fish_kvcache_attn_backend()` returns whether the Triton path installed, and the SlowAR model includes the plugin in `get_omni_plugins()` only on success; the runner-side check then stays a plain empty-list iteration with no availability logic.

**Review by linyueqian — COMMENTED** [review]
One piece of forward-looking advice from the TTS side (context: hsliu's cleanup ask in #4855 about model-specific logic living in worker/model_runner). The `OmniModelStatePlugin` + `get_omni_plugins()` design here is exactly the right mechanism, and worker_v2 is currently clean of the model-specific attention wiring that grew in the v1 runner. The one case that will test this cleanliness first is Fish Speech's kvcache backend, so it is worth deciding now how it migrates; details inline.

End-state note for honesty: the reason fish needs runner cooperation at all is that `fish_kvcache_backend` wraps module forwards instead of implementing a vLLM `AttentionBackend`. If it is ever rewritten as a real backend, the hooks below evaporate. That is out of scope for this PR; the plugin hooks are containment so the v2 runner never has to know a model's name.
